### PR TITLE
Awschristou/change notification

### PR DESCRIPTION
**Pull request by awschristou**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

General Feature:
* A reusable notify-on-startup system. Notices can be registered as extension points, containing their own logic around if a notice needs to be shown.

Concrete Usage:
* A notice is shown on startup for people using IDE versions older than 2019.2. The notice indicates that an upcoming toolkit release will require 2019.2 or newer.

Not in this Change:
* A Toolkit Settings facility to reset all silenced notifications

image

This is what it looks like when more than one notice appears
image

## Testing

Using the notification about needing 2019.2 in a future release...
* Ran IntelliJ 2019.2, notification did not appear.
* Ran IntelliJ 2019.1, saw notification appear.
  * Pressing "Don't show again" and restarting the IDE resulted in the notification not appearing again

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: resources/resources/software/aws/toolkits/resources/localized_messages.properties [hunk 1]
```diff
@@ -308,3 +308,7 @@ aws.settings.executables.find.description=Select path to {0} executable
 aws.settings.executables.cannot_determine_version=Cannot determine version of {0}: {1}
 aws.settings.executables.executable_invalid=Invalid {0} executable: {1}
 aws.settings.executables.resolution_exception=Exception occurred attempting to resolve {0} executable: {1}
+
+notice.suppress=Don't show this notification again
```
Comment (kiiadi) [inline resources/resources/software/aws/toolkits/resources/localized_messages.properties]: I'd shorten this to "Don't show this again"

File: resources/resources/software/aws/toolkits/resources/localized_messages.properties [hunk 2]
```diff
@@ -308,3 +308,7 @@ aws.settings.executables.find.description=Select path to {0} executable
...
+notice.title.jetbrains.minimum.version.2019.2=AWS Toolkit Requirements Notice
+notice.message.jetbrains.minimum.version.2019.2=An upcoming release of the AWS Toolkit will require version 2019.2 or newer of the JetBrains IDE.
```
Comment (kiiadi) [inline resources/resources/software/aws/toolkits/resources/localized_messages.properties]: Can we get smart about the IDE name? I think it would be clearer to customers if we said something like "require version 2019.2 of IntelliJ" /PyCharm/etc; should be able to use `ApplicationNamesInfo.getInstance().fullProductName`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeManager.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    private fun notify(notice: NoticeType, project: Project) {
+        val notification = notificationGroup.createNotification(
+            notice.getNoticeContents().title,
+            notice.getNoticeContents().message,
+            NotificationType.INFORMATION,
+            null
+        )
+
+        notification.addAction(
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeManager.kt]: like it!

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            }
+        }
+
+        return true
+    }
+
+    override fun getNoticeContents(): NoticeContents = noticeContents
+}
+
+data class NoticeContents(var title: String, var message: String)
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: can these be `val` ? or are they persisted somewhere?

Comment (awschristou) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: These are sent to the Notification balloon for display to the user so they are required. They aren't persisted.

**Comment by awschristou** [comment]
> Do you think we should have a standard "title" for all AWS Toolkit notices? Something like "AWS Toolkit Notification"
> 

I think there's a greater chance of different notice reasons. For example, this would be a good use of the "first time" notification to enable/disable telemetry settings. A more relevant title seems appropriate.

File: resources/resources/software/aws/toolkits/resources/localized_messages.properties (hunk 2 above)
Comment (awschristou) [reply resources/resources/software/aws/toolkits/resources/localized_messages.properties]: I like this suggestion, I'm taking a look

Comment (awschristou) [reply resources/resources/software/aws/toolkits/resources/localized_messages.properties]: Coming soon:
image

Comment (awschristou) [reply resources/resources/software/aws/toolkits/resources/localized_messages.properties]: image

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt (hunk 4 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Ok let's make them val then since it's immutable

Comment (awschristou) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Oops, `var` vs `val` was an oversight. I mis-read your initial question as `val?`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeManager.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            }
+
+            true
+        }
+
+    override fun notify(notices: List<NoticeType>, project: Project) {
+        notices.forEach { notify(it, project) }
+    }
+
+    private fun notify(notice: NoticeType, project: Project) {
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeManager.kt]: also take a look at `NotificationUtils`, which already has some of this logic

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.notification
+
+import com.intellij.openapi.application.ApplicationInfo
+import com.intellij.openapi.application.ApplicationNamesInfo
+import com.intellij.openapi.extensions.ExtensionPointName
+import software.aws.toolkits.resources.message
+
+interface NoticeType {
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Should we allow for additional actions? For instance, for the example case, you may want a link to upgrade the IDE.

Comment (awschristou) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: That's a good thought. I would consider this out of scope for this change, but `NoticeType` could certainly expose something like `getActions` which the Manager could add to the notification that is produced.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeManager.kt (hunk 5 above)
Comment (awschristou) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeManager.kt]: I decided to leave this as is (for now) so that I have the notification object for use with the `notification.hideBalloon()` call

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt (hunk 6 above)
Comment (bryceitoc9) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: I'm on the fence about considering this out-of-scope for the change, since the task is about writing the framework for these changes as well.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    override fun isNotificationRequired(): Boolean {
+        val appInfo = ApplicationInfo.getInstance()
+        val majorVersion = appInfo.majorVersion.toIntOrNull()
+        val minorVersion = appInfo.minorVersion.toIntOrNull()
+
+        majorVersion?.let {
+            minorVersion?.let {
+                return majorVersion < 2019 || (majorVersion == 2019 && minorVersion < 2)
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Will updating the min version involve upping these numbers? Can we pull from the plugin configuration instead?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    companion object {
+        val EP_NAME = ExtensionPointName<NoticeType>("aws.toolkit.notice")
+
+        internal fun notices(): List<NoticeType> = EP_NAME.extensions.toList()
+    }
+}
+
+class JetBrainsMinimumVersionChange : NoticeType {
+    override val id: String = "JetBrainsMinimumVersion_192"
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Will any new notifications for newer versions require us to manually up this value? Not a huge fan of that personally; it adds a lot of moving parts.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    private val noticeContents = NoticeContents(
+        message("notice.title.jetbrains.minimum.version.2019.2"),
+        message("notice.message.jetbrains.minimum.version.2019.2", ApplicationNamesInfo.getInstance().fullProductName)
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Will updating the min version involve upping these numbers? Can we pull from the plugin configuration instead?

(also goes for the line above)

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    fun getNoticeContents(): NoticeContents
+
+    companion object {
+        val EP_NAME = ExtensionPointName<NoticeType>("aws.toolkit.notice")
+
+        internal fun notices(): List<NoticeType> = EP_NAME.extensions.toList()
+    }
+}
+
+class JetBrainsMinimumVersionChange : NoticeType {
```
Comment (bryceitoc9) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Move this to a separate file

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt (hunk 6 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: YAGNI - it'll be easy to add in a backwards compatible way in the future - I tend to agree with @awschristou that it's not a requirement for this change. We've done as much of the generic version as we needed for our requirement and left it open for other notices to extend in the future.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt (hunk 7 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: Problem is we want the notice before we do the bump.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt (hunk 10 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/notification/NoticeType.kt]: agree
